js/src/evaluation/_runner.ts

```diff
@@ -0,0 +1,932 @@
+      }
+      return _resolveData(this._data, { client: this.client });
+    }
+    return this._examples;
+  }
+
+  get datasetId(): Promise<string> {
+    if (!this._experiment || !this._experiment.reference_dataset_id) {
+      const examplesIterator = this.examples[Symbol.asyncIterator]();
+      return examplesIterator.next().then((result) => {
+        if (result.done) {
+          throw new Error("No examples found in the dataset.");
+        }
+        return result.value.dataset_id;
+      });
+    }
+    return Promise.resolve(this._experiment.reference_dataset_id);
+  }
+
+  get evaluationResults(): AsyncGenerator<EvaluationResults> {
+    if (this._evaluationResults === undefined) {
+      return async function* (this: _ExperimentManager) {
+        for (const _ of this._examplesArray ?? []) {
+          yield { results: [] };
+        }
+      }.call(this);
+    } else {
+      return this._evaluationResults;
+    }
+  }
+
+  get runs(): AsyncGenerator<Run> {
+    if (this._runsArray && this._runsArray.length > 0) {
+      throw new Error("Runs already provided as an array.");
+    }
+    if (this._runs === undefined) {
+      throw new Error(
+        "Runs not provided in this experiment. Please predict first."
+      );
+    } else {
+      return this._runs;
+    }
+  }
+
+  async start(): Promise<_ExperimentManager> {
+    if (!this._examplesArray) {
+      this._examplesArray = [];
+    }
+    for await (const example of this.examples) {
+      this._examplesArray.push(example);
+    }
+    const firstExample = this._examplesArray[0];
+    const project = await this._getProject(firstExample);
+    this._printExperimentStart();
+    return new _ExperimentManager({
+      _examplesArray: this._examplesArray,
+      experiment: project,
+      metadata: this._metadata,
+      client: this.client,
+      evaluationResults: this._evaluationResults,
+      summaryResults: this._summaryResults,
+    });
+  }
+
+  async withPredictions(
+    target: TargetT,
+    options?: {
+      maxConcurrency?: number;
+    }
+  ): Promise<_ExperimentManager> {
+    const experimentResults = this._predict(target, options);
+    return new _ExperimentManager({
+      _examplesArray: this._examplesArray,
+      experiment: this._experiment,
+      metadata: this._metadata,
+      client: this.client,
+      runs: (async function* (): AsyncGenerator<Run> {
+        for await (const pred of experimentResults) {
+          yield pred.run;
+        }
+      })(),
+    });
+  }
+
+  async withEvaluators(
+    evaluators: Array<EvaluatorT | RunEvaluator>,
+    options?: {
+      maxConcurrency?: number;
+    }
+  ): Promise<_ExperimentManager> {
+    const resolvedEvaluators = _resolveEvaluators(evaluators);
+    const experimentResults = this._score(resolvedEvaluators, options);
+
+    const results: AsyncGenerator<ExperimentResultRow>[] = [];
+    for await (const item of atee<ExperimentResultRow>(experimentResults)) {
```

Comment:
Doesn't this consume the generator anyway? Don't think we need the `tee` operation then?

Reply:
We return two generators here, which both use this original generator. We could get rid of it in favor of consuming `experimentResults` inside `withEvaluators` and returning the data instead of new generators.

Reply:
I'm less famliar with JS but looks like `atee` itself is sync right? Why do we need to do `for await (const...)`, etc. rather than just `const [r1, r2] = atee<ExperimentResultRow>(experimentResults)`

Reply:
Ah this is correct. The prev implementation I added was async which is why it was like that. Pushed up a fix